### PR TITLE
rustc_resolve: don't treat uniform_paths canaries as ambiguities unless they resolve to distinct Def's.

### PR DESCRIPTION
**Pull request by eddyb**
In particular, this allows this pattern that @cramertj mentioned in https://github.com/rust-lang/rust/issues/53130#issuecomment-420848814: 
```rust
use log::{debug, log};
fn main() {
    use log::{debug, log};
    debug!(...);
}
```
The canaries for the inner `use log::...;`, *in the macro namespace*, see the `log` macro imported at the module scope, and the (same) `log` macro, imported in the block scope inside `main`.

Previously, these two possible (macro namspace) `log` resolutions would be considered ambiguous (from a forwards-compat standpoint, where we might make imports aware of block scopes).

With this PR, such a case is allowed *if and only if* all the possible resolutions refer to the same definition (more specifically, because the *same* `log` macro is being imported twice).
This condition subsumes previous (weaker) checks like #54005 and the second commit of #54011.

Only the last commit is the main change, the other two are cleanups.

r? @petrochenkov cc @Centril @joshtriplett

### CODE REVIEW

**Comment by petrochenkov** [comment]
>The canaries for the inner use log::...;, in the macro namespace, see the log macro imported at the module scope, and the (same) log macro, imported in the block scope inside main.

Do you mean one `use log as _;` conflicts with another `use log as _;`?
I don't think this should happen - canaries for `use prefix::...` need only be created in type namespace.
Canaries in other namespaces are only necessary for single segment impors (`use log;`).

**Comment by petrochenkov** [comment]
Ok, that's an orthogonal issue, regardless of that resolutions with same `Def` are acceptable.
@bors r+

File: src/librustc_resolve/resolve_imports.rs [hunk 1]
```diff
@@ -720,77 +725,72 @@ impl<'a, 'b:'a, 'c: 'b> ImportResolver<'a, 'b, 'c> {
...
-                    true
-                };
-                let module_scope = results[ns].module_scope.filter(result_filter);
-                let block_scopes = || {
-                    results[ns].block_scopes.iter().cloned().filter(result_filter)
-                };
+            // Currently imports can't resolve in non-module scopes,
+            // we only have canaries in them for future-proofing.
+            if external_crate.is_none() && results.module_scope.is_none() {
+                return;
```
Comment (eddyb) [inline src/librustc_resolve/resolve_imports.rs]: Oops, this is now a `for` loop so I need to `continue` instead!

Pretty sure this caused #54253, will fix in #54116.

**Pull request merged**